flytekit-java/src/main/java/org/flyte/flytekit/SdkBindingData.java

```diff
@@ -38,215 +37,203 @@
+    return ofPrimitive(Primitive::ofStringValue, value);
   }
 
-  public static SdkBindingData ofBoolean(boolean value) {
-    return ofPrimitive(Primitive.ofBooleanValue(value));
+  public static SdkBindingData<Boolean> ofBoolean(boolean value) {
+    return ofPrimitive(Primitive::ofBooleanValue, value);
   }
 
-  public static SdkBindingData ofDatetime(int year, int month, int day) {
+  public static SdkBindingData<Instant> ofDatetime(int year, int month, int day) {
     Instant instant = LocalDate.of(year, month, day).atStartOfDay().toInstant(ZoneOffset.UTC);
     return ofDatetime(instant);
   }
 
-  public static SdkBindingData ofDatetime(Instant value) {
-    return ofPrimitive(Primitive.ofDatetime(value));
-  }
-
-  public static SdkBindingData ofDuration(Duration value) {
-    return ofPrimitive(Primitive.ofDuration(value));
+  public static SdkBindingData<Instant> ofDatetime(Instant value) {
+    return ofPrimitive(Primitive::ofDatetime, value);
   }
 
-  public static SdkBindingData ofOutputReference(String nodeId, String nodeVar, LiteralType type) {
-    BindingData idl =
-        BindingData.ofOutputReference(
-            OutputReference.builder().nodeId(nodeId).var(nodeVar).build());
-    return create(idl, type);
+  public static SdkBindingData<Duration> ofDuration(Duration value) {
+    return ofPrimitive(Primitive::ofDuration, value);
   }
 
-  public static SdkBindingData ofPrimitive(Primitive primitive) {
+  private static <T> SdkBindingData<T> ofPrimitive(Function<T, Primitive> toPrimitive, T value) {
+    Primitive primitive = toPrimitive.apply(value);
     BindingData bindingData = BindingData.ofScalar(Scalar.ofPrimitive(primitive));
     LiteralType literalType = LiteralType.ofSimpleType(getSimpleType(primitive.kind()));
 
-    return create(bindingData, literalType);
-  }
-
-  public static SdkBindingData ofStruct(SdkStruct struct) {
-    BindingData bindingData = BindingData.ofScalar(Scalar.ofGeneric(struct.struct()));
-    LiteralType literalType = LiteralType.ofSimpleType(SimpleType.STRUCT);
-
-    return create(bindingData, literalType);
+    return create(bindingData, literalType, value);
   }
 
-  private static SimpleType getSimpleType(Primitive.Kind kind) {
-    switch (kind) {
-      case INTEGER_VALUE:
-        return SimpleType.INTEGER;
-      case FLOAT_VALUE:
-        return SimpleType.FLOAT;
-      case STRING_VALUE:
-        return SimpleType.STRING;
-      case BOOLEAN_VALUE:
-        return SimpleType.BOOLEAN;
-      case DATETIME:
-        return SimpleType.DATETIME;
-      case DURATION:
-        return SimpleType.DURATION;
-    }
-
-    throw new AssertionError("Unexpected Primitive.Kind: " + kind);
+  public static <T> SdkBindingData<List<T>> ofCollection(
+      List<T> collection, Function<T, SdkBindingData<T>> function) {
+    return SdkBindingData.ofBindingCollection(
+        collection.stream().map(function).collect(Collectors.toList()));
```

Comment:
This API feels a little off to me. When is it expected to be used over the (to me) simpler `ofBindingCollection`? I can see for `ofBindingCollection` that one the user needs to wrap their list elements themselves but they likely know well enough how to operate on lists in Java. For `ofCollection` they would need to first understand what the `function` parameter does (could be documented I guess) and then know what `SdkBindingData.of*` function to put there (this they would need to know to wrap the values themselves too though). Perhaps renaming the `function` parameter to `mapper` or `transform` or something could help a bit.

```suggestion
  public static <T> SdkBindingData<List<T>> ofCollection(
      List<T> collection, Function<T, SdkBindingData<T>> mapper) {
    return SdkBindingData.ofBindingCollection(
        collection.stream().map(mapper).collect(Collectors.toList()));
```

Reply:
For empty lists, I'm not sure what is worse/best:

```java
SdkBindingData.ofCollection(List.<String>of(), SdkBindingData::ofString);
SdkBindingData.ofBindingCollection(List.<SdkBindingData<String>>of());
```

Reply:
We can also reconsider adding some helper methods for common types like ofStringCollection(...), ofBooleanCollection(...), etc. (but I would say that would be for a future PR)

Reply:
@pablocasares added these helper methods.